### PR TITLE
feature/lisasiliu/edit-programs-form-backend-route

### LINKED ISSUE #48: [Edit Program Information] - Create a backend route that allows you to edit a program

### PR DESCRIPTION
**Pull request by lisasiliu**
## Tracking Info

Resolves [#21](https://github.com/orgs/TritonSE/projects/9?pane=issue&itemId=52062799)

## Changes

Added edit (patch) route, validator for updates, updateForm

## Testing

Tested with PATCH route in Postman. Tested incorrect ID, missing ID in path, and IDs not in database.

## Confirmation of Change

Request bodies use following structure.

Example with object ID 660afbb0e7a023463d3586f3.
{
    "_id": "660afbb0e7a023463d3586f3",
    "name": "A New Program",
    "abbreviation": "ANP",
    "type": "regular",
    "daysOfWeek": ["M", "W", "F"],
    "startDate": "2022-09-10T11:00:55Z",
    "endDate": "2023-08-10T11:00:55Z",
    "color": "#FFFFFF"
}
Postman path: PATCH http://localhost:4000/api/program/660afbb0e7a023463d3586f3

To test, can first POST ([x](https://github.com/TritonSE/PIA-Program-Manager/pull/35)) to the database, then use the generated object ID to use in PATCH request.

Successful PATCH:
success1

Failed PATCH (ID in url does not match ID in request body)
fail1

Failed PATCH (ID does not exist in MongoDB database):
fail2

### CODE REVIEW

File: backend/src/controllers/program.ts [hunk 1]
```diff
@@ -28,3 +30,29 @@ export const createForm: RequestHandler = async (req, res, next) => {
...
+
+export const updateForm: RequestHandler = async (req, res, next) => {
+  const errors = validationResult(req);
+  try {
+    validationErrorParser(errors);
+
+    const programId = req.params.id;
+    const programData = req.body as Program;
+
+    if (programId !== programData._id) {
```
Comment (adhi0331) [inline backend/src/controllers/program.ts]: Is there a reason why we are passing in the program id in the request body? We already get it from req.params.id. Besides we don't want users to be able to change a program id or any id's for that matter.

File: backend/src/controllers/program.ts [hunk 2]
```diff
@@ -28,3 +30,29 @@ export const createForm: RequestHandler = async (req, res, next) => {
     next(error);
   }
 };
+
+export const updateForm: RequestHandler = async (req, res, next) => {
```
Comment (adhi0331) [inline backend/src/controllers/program.ts]: Let's rename the function to updateProgram. Also update createForm to createProgram

File: backend/src/validators/program.ts [hunk 3]
```diff
@@ -112,3 +119,14 @@ export const createForm = [
   makeEndDateValidator(),
   makeColorValidator(),
 ];
+
+export const updateForm = [
```
Comment (adhi0331) [inline backend/src/validators/program.ts]: Rename this to updateProgram and createProgram

**Review by adhi0331 — COMMENTED** [review]
Looks good to me and the route works. If you could answer the one question I have then I think this is good to merge!

**Comment by lisasiliu** [comment]
made changes, only issue now is when ID is invalid, the error is "unknown error"
image

**Comment by lisasiliu** [comment]
(not sure why the lint failed, i didn't change anything in user.ts T_T)

**Review by mraysu — APPROVED** [review]
Looks clean! I got a head start on the frontend routing and I think I \*should\* have no problems merging this in. There are a few changes that will need to be made in order to match the V2 program schema but other than that it looks good.

**Review by adhi0331 — APPROVED** [review]
LGTM! 🚀

**Pull request merged**